test/common/integration/xds_integration_test.cc

```diff
@@ -0,0 +1,188 @@
+    bootstrap.mutable_static_resources()->add_clusters()->MergeFrom(
+        createSingleEndpointClusterConfig("base_h2"));
+    bootstrap.mutable_static_resources()->add_clusters()->MergeFrom(
+        createSingleEndpointClusterConfig("sds_cluster.lyft.com"));
+  });
+
+  // xDS upstream is created separately in the test infra, and there's only one non-xDS cluster.
+  setUpstreamCount(1);
+
+  // Add the Admin config.
+  config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    bootstrap.mutable_admin()->MergeFrom(adminConfig());
+  });
+  admin_filename_ = TestEnvironment::temporaryPath("admin_address.txt");
+  setAdminAddressPathForTests(admin_filename_);
+}
+
+Network::Address::IpVersion XdsIntegrationTest::ipVersion() const {
+  return std::get<0>(GetParam());
+}
+
+Grpc::ClientType XdsIntegrationTest::clientType() const { return std::get<1>(GetParam()); }
+
+Grpc::SotwOrDelta XdsIntegrationTest::sotwOrDelta() const { return std::get<2>(GetParam()); }
+
+std::string XdsIntegrationTest::loopbackAddr() const {
+  if (ipVersion() == Network::Address::IpVersion::v6) {
+    return "::1";
+  }
+  return "127.0.0.1";
+}
+
+void XdsIntegrationTest::SetUp() {
+  // TODO(abeyad): Add paramaterized tests for HTTP1, HTTP2, and HTTP3.
+  setUpstreamProtocol(Http::CodecType::HTTP2);
+}
+
+void XdsIntegrationTest::TearDown() {
+  cleanup();
+  BaseClientIntegrationTest::TearDown();
+}
+
+void XdsIntegrationTest::createEnvoy() {
+  BaseClientIntegrationTest::createEnvoy();
+  std::string admin_str = TestEnvironment::readFileToStringForTest(admin_filename_);
+  auto addr = Network::Utility::parseInternetAddressAndPort(admin_str);
+  registerPort("admin", addr->ip()->port());
+  if (on_server_init_function_) {
+    on_server_init_function_();
+  }
+}
+
+void XdsIntegrationTest::initializeXdsStream() {
+  createXdsConnection();
+  AssertionResult result =
+      xds_connection_->waitForNewStream(*BaseIntegrationTest::dispatcher_, xds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  xds_stream_->startGrpcStream();
+}
+
+std::string XdsIntegrationTest::getRuntimeKey(const std::string& key) {
+  auto response = IntegrationUtil::makeSingleRequest(
+      lookupPort("admin"), "GET", "/runtime?format=json", "", Http::CodecType::HTTP2, version_);
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+  Json::ObjectSharedPtr loader = TestEnvironment::jsonLoadFromString(response->body());
+  auto entries = loader->getObject("entries");
+  if (entries->hasObject(key)) {
+    return entries->getObject(key)->getString("final_value");
+  }
+  return "";
+}
+
+uint64_t XdsIntegrationTest::getCounterValue(const std::string& counter) {
+  auto response = IntegrationUtil::makeSingleRequest(lookupPort("admin"), "GET", "/stats?usedonly",
+                                                     "", Http::CodecType::HTTP2, version_);
+  EXPECT_TRUE(response->complete());
```

Comment:
optional follow up I think it'd still be nice to get this happening in pulse client rather than custom in this test so it's reusable

Reply:
ACK, I added a TODO in the code for this.